pkg/distributor/distributor.go

```diff
@@ -1223,12 +1226,27 @@ func (d *Distributor) prePushValidationMiddleware(next PushFunc) PushFunc {
 			skipLabelCountValidation := d.cfg.SkipLabelCountValidation || req.GetSkipLabelCountValidation()
 
 			// Note that validateSeries may drop some data in ts.
+			rawSamples := len(ts.Samples)
+			rawHistograms := len(ts.Histograms)
 			shouldRemove, validationErr := d.validateSeries(now, &req.Timeseries[tsIdx], userID, group, skipLabelValidation, skipLabelCountValidation, minExemplarTS, maxExemplarTS)
 
 			if countDroppedNativeHistograms {
 				droppedNativeHistograms += len(ts.Histograms)
 			}
 
+			dedupedSamplesAndHistograms := (rawSamples - len(ts.Samples)) + (rawHistograms - len(ts.Histograms))
+			if dedupedSamplesAndHistograms > 0 {
+				for _, l := range ts.Labels {
+					if l.Name != labels.MetricName {
+						continue
+					}
+					if _, exists := dedupedPerMetric[l.Value]; exists || len(dedupedPerMetric) < maxTrace {
+						dedupedPerMetric[l.Value] += dedupedSamplesAndHistograms
+						break
```

Comment:
I don't follow this condition. It breaks the loop when a series's name exists in the set, or the length of the set exceeds the max. Where does the code insert something into the set?

Reply:
Line 1244.

Reply:
Line 1244